src/kontrol/prove.py

```diff
@@ -757,9 +757,14 @@ def _update_cterm_from_node(cterm: CTerm, node: KCFG.Node, config_type: ConfigTy
     return new_init_cterm
 
 
-def deployment_state_to_account_cells(deployment_state_entries: Iterable[DeploymentStateEntry]) -> list[KApply]:
-    accounts = _process_deployment_state(deployment_state_entries)
-    address_list = accounts.keys()
+def recorded_state_to_account_cells(
+    recorded_state_entries: Iterable[StateDiffEntry] | Iterable[StateDumpEntry],
+) -> list[KApply]:
+    if isinstance(list(recorded_state_entries)[0], StateDiffEntry):
+        accounts = _process_state_diff(recorded_state_entries)
```

Comment:
Here `mypy` complains about `recorded_state_entries` having the wrong type:
```
error: Argument 1 to "_process_state_diff" has incompatible type "Iterable[StateDiffEntry] | Iterable[StateDumpEntry]"; expected "Iterable[StateDiffEntry]"  [arg-type]
```
But I don't know of a better way to pass the `recorded_state_entries` to the correct function. Any insight will be appreciated!

Reply:
Hm, interesting! I thought changing 
```python
if isinstance(list(recorded_state_entries)[0], StateDiffEntry):
```
to 
```py
if all(isinstance(entry, StateDiffEntry) for entry in recorded_state_entries):
```
would prevent it from complaining, but I guess it's still not sufficient for `mypy` to infer the types through static analysis. Very interested in what the correct way to address it is too!

Reply:
`TypeGuard` was our answer :)